util/confutil/config.go

```diff
@@ -1,39 +1,140 @@
 package confutil
 
 import (
+	"crypto/rand"
+	"encoding/hex"
 	"os"
-	"path"
 	"path/filepath"
+	"strings"
+	"sync"
 
 	"github.com/docker/cli/cli/command"
+	"github.com/docker/docker/pkg/ioutils"
 	"github.com/pelletier/go-toml"
 	"github.com/pkg/errors"
-	"github.com/sirupsen/logrus"
 )
 
-// ConfigDir will look for correct configuration store path;
-// if `$BUILDX_CONFIG` is set - use it, otherwise use parent directory
-// of Docker config file (i.e. `${DOCKER_CONFIG}/buildx`)
-func ConfigDir(dockerCli command.Cli) string {
-	if buildxConfig := os.Getenv("BUILDX_CONFIG"); buildxConfig != "" {
-		logrus.Debugf("using config store %q based in \"$BUILDX_CONFIG\" environment variable", buildxConfig)
-		return buildxConfig
+var (
+	sudoerUID = -1
+	sudoerGID = -1
+)
+
+type Config struct {
+	dir string
+}
+
+type ConfigOption func(*configOptions)
+
+type configOptions struct {
+	dir string
+}
+
+func WithDir(dir string) ConfigOption {
+	return func(o *configOptions) {
+		o.dir = dir
 	}
+}
 
-	buildxConfig := filepath.Join(filepath.Dir(dockerCli.ConfigFile().Filename), "buildx")
-	logrus.Debugf("using default config store %q", buildxConfig)
-	return buildxConfig
+func NewConfig(dockerCli command.Cli, opts ...ConfigOption) *Config {
+	co := configOptions{}
+	for _, opt := range opts {
+		opt(&co)
+	}
+	configDir := co.dir
+	if configDir == "" {
+		configDir = os.Getenv("BUILDX_CONFIG")
+		if configDir == "" {
+			configDir = filepath.Join(filepath.Dir(dockerCli.ConfigFile().Filename), "buildx")
+		}
+	}
+	return &Config{
+		dir: configDir,
+	}
 }
 
-// DefaultConfigFile returns the default BuildKit configuration file path
-func DefaultConfigFile(dockerCli command.Cli) (string, bool) {
-	f := path.Join(ConfigDir(dockerCli), "buildkitd.default.toml")
+// Dir will look for correct configuration store path;
+// if `$BUILDX_CONFIG` is set - use it, otherwise use parent directory
+// of Docker config file (i.e. `${DOCKER_CONFIG}/buildx`)
+func (c *Config) Dir() string {
+	return c.dir
+}
+
+// BuildKitConfigFile returns the default BuildKit configuration file path
+func (c *Config) BuildKitConfigFile() (string, bool) {
+	f := filepath.Join(c.dir, "buildkitd.default.toml")
 	if _, err := os.Stat(f); err == nil {
 		return f, true
 	}
 	return "", false
 }
 
+// MkdirAll creates a directory and all necessary parents within the config dir
+func (c *Config) MkdirAll(dir string, perm os.FileMode) error {
+	d := filepath.Join(c.dir, dir)
+	if err := os.MkdirAll(d, perm); err != nil {
+		return err
+	}
+	if sudoerUID != -1 && sudoerGID != -1 {
+		// apply chown to each directory level
+		parts := strings.Split(dir, string(filepath.Separator))
+		for i := range parts {
+			if err := os.Chown(filepath.Join(c.dir, filepath.Join(parts[:i+1]...)), sudoerUID, sudoerGID); err != nil {
```

Comment:
Shouldn't this only run if `MkdirAll` actually created the directory component?

Reply:
Hum maybe but that would fix currently broken permissions on folders for some users. We could check if last directory level has right perms and skip this iteration?

Reply:
I don't think last component makes a difference as `MkdirAll` should have same behavior for all components.

Maybe run `stat` in here and if the owner is root but the conditions are correct for sudo then chown.

Reply:
As this logic already exists in `fsutil` in `copy` package, I switched to `fs.MkdirAll`: https://github.com/tonistiigi/fsutil/blob/3f140a1299b0928e425ca320d5b8eb396eee0d66/copy/mkdir.go#L10